binding/react/src/use_leopard.ts

```diff
@@ -0,0 +1,254 @@
+          setError(null);
+        }
+      } catch (e: any) {
+        setError(e);
+      }
+    },
+    []
+  );
+
+  const process = useCallback(async (pcm: Int16Array): Promise<void> => {
+    try {
+      if (!leopardRef.current) {
+        setError(
+          new Error('Leopard has not been initialized or has been released')
+        );
+        return;
+      }
+
+      const processResult = await leopardRef.current.process(
+        pcm,
+        processOptionsRef.current
+      );
+      setResult(processResult);
+    } catch (e: any) {
+      setError(e);
+    }
+  }, []);
+
+  const processFile = useCallback(async (audioFile: File): Promise<void> => {
+    if (!leopardRef.current) {
+      setError(
+        new Error('Leopard has not been initialized or has been released')
+      );
+      return new Promise(resolve => resolve());
+    }
+
+    // @ts-ignore
+    const audioContext = new (window.AudioContext || window.webKitAudioContext)(
+      { sampleRate: leopardRef.current.sampleRate }
+    );
+
+    const readAudioFile = (
+      file: File,
+      callback: (audioBuffer: AudioBuffer) => Promise<void>
+    ): void => {
+      const reader = new FileReader();
+
+      reader.onload = (): void => {
+        const wavBytes = reader.result;
+        audioContext.decodeAudioData(wavBytes as ArrayBuffer, callback);
+      };
+
+      reader.readAsArrayBuffer(file);
+    };
+
+    return new Promise<void>(resolve => {
+      readAudioFile(audioFile, async (audioBuffer: AudioBuffer) => {
+        const f32PCM = audioBuffer.getChannelData(0);
+        const i16PCM = new Int16Array(f32PCM.length);
+        const INT16_MAX = 32767;
+        const INT16_MIN = -32768;
+        i16PCM.set(
+          f32PCM.map(f => {
+            let i = Math.trunc(f * INT16_MAX);
+            if (f > INT16_MAX) i = INT16_MAX;
+            if (f < INT16_MIN) i = INT16_MIN;
+            return i;
+          })
+        );
+
+        await process(i16PCM);
+        resolve();
+      });
+    });
+  }, []);
+
+  const stopRecording = useCallback(async (): Promise<void> => {
+    try {
+      if (timerRef.current) {
+        clearInterval(timerRef.current);
+
+        await WebVoiceProcessor.unsubscribe(recorderEngineRef.current);
+        setIsRecording(false);
+
+        const frames = new Int16Array(audioDataRef.current.length * 512);
+        for (let i = 0; i < audioDataRef.current.length; i++) {
+          frames.set(audioDataRef.current[i], i * 512);
+        }
+        await process(frames);
+
+        audioDataRef.current = [];
+      }
+    } catch (e: any) {
+      setError(e);
+    }
+  }, []);
+
+  const startRecording = useCallback(
+    async (maxRecordingSec = DEFAULT_MAX_RECORDING_SEC): Promise<void> => {
+      if (!leopardRef.current) {
+        setError(
+          new Error('Leopard has not been initialized or has been released')
+        );
+        return;
+      }
+
+      setError(null);
+      audioDataRef.current = [];
```

Comment:
Unless I'm missing something, it looks as though we aren't checking if we're currently recording or not before we try to start recording again. Generally, you'd want to check if we're recording and then either return right away (because there's nothing to do) or throw an error if we want to be more harsh.